### PR TITLE
Remove old unqualified lookup implementation

### PR DESCRIPTION
**Pull request by slavapestov**
The only remaining blocker to removing the old unqualified lookup implementation were a few places where we would perform an unqualified lookup from a non-top-level DeclContext, without a SourceLoc. This is no longer supported, since ASTScope always uses a SourceLoc to locate the innermost scope. First, this PR refactors these lookups to either start from a top-level context if there's no change in functionality, or perform a qualified lookup into a type if needed. With that out of the way, we can stick the right assertion in UnqualifiedLookup and remove the old code.

The next steps here are removing parse-time local variable lookup, since ASTScope can always find declarations in local scope, and refactoring LookupVisibleDecls to consult ASTScope instead of walking the AST directly. Once that is done, we'll finally have one canonical source of truth for unqualified lookups, concluding the ASTScope journey.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 1]
```diff
@@ -486,38 +486,66 @@ template <typename FnTy>
...
+
+  auto type = DC->getDeclaredInterfaceType();
+  if (!type->is<ErrorType>()) {
+    DC->lookupQualified(type, DeclNameRef(VD->getBaseName()),
+                        NL_QualifiedDefault, results);
+  }
+
+  SmallVector<ValueDecl *, 8> RelatedDecls;
+  for (auto result : results) {
+    if (result->getAttrs().isUnavailable(ctx))
```
Comment (davidungar) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:537]: This whole function could be easier to understand with either a comment describing the overall strategy, or by breaking it up.

**Review by davidungar — APPROVED** [review]
Great to see this finally ripped out!! There is one place where a maintainer would have to spend a little time to figure out what's going on. Either breaking up that function or adding a comment about the overall flow could help.

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp (hunk 1 above)
Comment (slavapestov) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:537]: You're right. I completely refactored this function, will push a new version shortly.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp (hunk 1 above)
Comment (davidungar) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp]: Thanks!

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 2]
```diff
@@ -484,40 +485,48 @@ void SwiftLangSupport::printFullyAnnotatedSynthesizedDeclaration(
...
       continue;
 
-    if (RelatedVD != VD) {
-      ++NamesSeen[RelatedVD->getName()];
+    if (result != VD) {
+      ++NamesSeen[result->getName()];
```
Comment (davidungar) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:522]: Looking pretty good--easy to understand.

**Pull request merged**